pkg/runner/job_executor.go

```diff
@@ -39,6 +39,11 @@ func newJobExecutor(info jobInfo, sf stepFactory, rc *RunContext) common.Executo
 	preSteps = append(preSteps, info.startContainer())
 
 	for i, stepModel := range infoSteps {
+		if stepModel == nil {
+			return func(ctx context.Context) error {
+				return fmt.Errorf("invalid Step %v: missing run or uses key", i)
+			}
+		}
 		if stepModel.ID == "" {
```

Comment:
This is the nil dereference, which this change aims to replace with a sane error. 
> I've created https://github.com/nektos/act/pull/1157 to fix this

The same happens with `if s.Run == "" && s.Uses == "" {`, since `s` can be nil based on the bugreport.
_This cannot happen in my usecase for act, since github doesn't sent such invalid steps_

Reply:
Please add a test for it then. We should be able to make sure it didn't happen again.

Reply:
Also what I've meant is that we should fix the issue as early as possible.
The best would be right after yaml parsing. Otherwise we will have to add the fix in different places across the codebase.

Reply:
The bigger fix would be removing the pointer from the Job.Steps `[]Step*` array to `[]Step`. Then it can never be nil, like in composite action steps. Seems like a large refactor, instead of a small patch without conflicts. You gave such conflicting PR's open.

Reply:
Fine for me as well to merge this patch, but let's add a test case please.

Reply:
Test case from description added, act is still pretty unstable.